src/query/intersection.rs

```diff
@@ -95,32 +100,39 @@ impl<TDocSet: DocSet, TOtherDocSet: DocSet> DocSet for Intersection<TDocSet, TOt
     fn advance(&mut self) -> DocId {
         let (left, right) = (&mut self.left, &mut self.right);
         let mut candidate = left.advance();
+        if candidate == TERMINATED {
+            return TERMINATED;
+        }
 
-        'outer: loop {
+        loop {
             // In the first part we look for a document in the intersection
             // of the two rarest `DocSet` in the intersection.
 
             loop {
-                let right_doc = right.seek(candidate);
-                candidate = left.seek(right_doc);
-                if candidate == right_doc {
+                if right.seek_exact(candidate) {
                     break;
                 }
+                // `left.advance().max(right.doc())` yielded a regression in the search game
+                // benchmark It may make sense in certain scenarios though.
+                candidate = left.advance();
```

Comment:
if it is small I would sleep better with the max(right.doc())

Reply:
The regression was not small, so it might be better to identify the cases in which `max(right.doc())` is useful and have a separate loop for them

Reply:
i can give you one adversarial case. 

intersection between two terms. 

The first term matches 0..1_000_000.
The second term is matching docs 1_000_000..10_000_000.

a more like scenario could be 
first term matches 0..1_000_000
second term is random.


This kind of problem actually happen in reality when document are ordered in a certain way.
For instance, eBay and Indeed are sorting by location.
We often sort by time, which is correlated in strange ways (for instance you can think of status:ERROR).

Reply:
Generally I think it would make sense for all queries that don't implement a specialized `seek_exact` to use the `max(right.doc())` loop. 
I don't think there's a way in rust to detect this?

For now we could just detect if the right docset is a `TermScorer` and then use the `max(right.doc())` variant. 

I created a issue here to add some missing benchmarks:
https://github.com/quickwit-oss/tantivy/issues/2544

Reply:
How large was the regression here? Do you have benchmarks?

Reply:
Actually I think the benchmark was bogus
1. We are not allowed to call `max(right.doc)` in the general case. 
2. If the candidate is coming from right, we would need to seek_exact on left.

It wasn't covered by a test then but it is now via proptest.